components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java

```diff
@@ -1570,6 +1570,40 @@ public static void validateOauthApplication(String consumerKey) throws InvalidAp
         }
     }
 
+    /**
+     * This API validate the oauth application access status.
+     * Check whether an application exits for given consumerKey and check whether user can login into.
+     *
+     * @param consumerKey clientId
+     * @throws InvalidApplicationClientException
+     */
+    public static void validateAppAccess(String consumerKey) throws InvalidApplicationClientException,
+            OAuthSystemException {
+
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(consumerKey);
```

Comment:
Let's check with a sample scenario whether the SP is loaded from the cache or not. If from DB, lets use a basic SP load method to get this property as the SP loading has multiple db calls.

Reply:
AFAIK the Cachelayer is at the DAO layer of the application mgt service and I dont think we need to handle the caching part here.

Reply:
Means this validation will be restricted for pwd and client_credentials grant. Ack. Will adhere.

Reply:
It was decided to move this back to the token endpoint.

Reply:
Not to handle caching, but to verify whether the result is fetched from a cache. If so we don't introduce a performance impact.